docs/CONTRIBUTING.md

```diff
@@ -0,0 +1,129 @@
+
+**Please note: We are not currently accepting Pull Requests to this repository.**
+
+## Development Philosophy
+
+This codebase aims to achieve the following goals:
+
+- Optimise for the principle of least surprise for developers
+- Aim to reduce churn and reduce the amount of different files touched during a change.
+- Reduce the likelyhood of bugs.
+
+We achieve these with the following:
+
+- **Consistency:** All areas of the codebase should behave in the same way and follow the same rules.
+- **Purpose:** Files, classes and methods should have clear single purposes.
+- **Test coverage:** We target as close to 100% test coverage of moving parts (models, commands, etc) and anticipated paths in user interactions (e.g. normal usage or error states).
+
+## Code Standards
+
+These code standards follow the principle of "strong opinions, loosely held".
+If something is in this file, we treat it as gospel - it should be adhered to strictly and everyone should respect it.
+However, we acknowledge that our standards might be incorrect and questioning them is sensible, so we welcome discussion of them.
+PRs to this document are welcome, and should be accompanied by changes to an example part of the codebase that would be affected.
+Final decisions are made by @iHiD but should be revisited periodically to test them against new experience and data.
+
+## API, SPI, and normal routes
+
+Our API (Application Programming Interface) is a public JSON API.
+It can be authenticated by either a session or by passing an authentication token.
+It should be used by View Components when they want to retrieve data (e.g. when filtering solutions).
+Non-trivial JSON should be returned via serializers.
+
+The SPI (Secure Programming Interface) is a private JSON API.
+It is used by other components in our architecture to pass messages and data around.
+It is not accessible via the Internet.
+
+All other "normal" routes redirect or render HTML.
+
+## Controllers
+
+Controllers should be "thin".
+Each action should do at most one thing, retrieve at most one thing, and render/redirect.
+To achieve this, controllers call out to Commands, which contain more complex functionality encapsualted in stand-alone procedures. This is know as the Command Pattern or the Interactor Pattern.
+
+## Serializers
+
+To ensure that data is represented in common ways, we use serializers.
+For example, the data for the mentoring workspace would be represented via a serializer.
+The HTML that renders the table as a React Component would use the serializer to generate the JSON for the initial React Component, and the API endpoint would use the same Serializer when the table's data is filtered.
+
+## Models
+
+Models are light-weight database wrappers.
+We treat models as highly-predictable, meaning that creating and updating models should not have side-effects.
+We therefore use `before_xxx` and `after-xxx` blocks sparsely.
```

Comment:
s/sparsely/sparingly

Reply:
@dougal Did you know that GitHub allows for comments to include suggestions? See https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request This makes it really easy for the PR submitter to apply changes. 

For an example what this looks like, see the suggestions made by kntsoriano on this PR.

Reply:
Ooh, fancy! TIL

Thanks @ErikSchierboom.

Reply:
My pleasure! I've made extensive use of that feature and I love it :D